datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -665,8 +670,30 @@ fn string_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType>
     }
 }
 
-/// Coercion rules for Binaries: the type that both lhs and rhs can be
-/// casted to for the purpose of a computation
+/// Coercion rules for binary (Binary/LargeBinary) to string (Utf8/LargeUtf8):
+/// If one argument is binary and the other is a string then coerce to string
+/// (e.g. for `like`)
+fn binary_to_string_coercion(
```

Comment:
this is the new rule, only used for `ilke`

I originally tried modifying `string_coercion` to also include `Binary --> String` coercion, however, that has the (bad) side effect of making a `binary = string` comparison into `cast(binary)= string` which is bad because:
1. The comparison is valid even if the binary column contains non UTF8 data (but the cast will fail in this case)
2. The cast from binary --> utf8 is not free (it has to check each value for utf8 correctness)